adala/utils/llm.py

```diff
@@ -0,0 +1,235 @@
+        raise ValueError("You must provide either `user_prompt` or `messages`.")
+
+    if not messages:
+        # get messages from user_prompt and system_prompt
+        messages = get_messages(user_prompt, system_prompt, instruction_first)
+    response = LLMResponse()
+
+    if response_model is None:
+        # unconstrained generation - return raw completion text and store it in `data` field: {"text": completion_text}
+        try:
+            completion = await litellm.acompletion(
+                model=model,
+                api_key=api_key,
+                messages=messages,
+                max_tokens=max_tokens,
+                temperature=temperature,
+                timeout=timeout,
+            )
+            completion_text = completion.choices[0].message.content
+            response.data['_completion_text'] = completion_text
+        except Exception as e:
+            response.adala_error = True
+            response.adala_message = type(e).__name__
+            # create a traceback string
+            response.adala_details = traceback.format_exc()
+        finally:
+            return response.model_dump(by_alias=True)
+
+    # constrained generation branch - use `response_model` to constrain the LLM response
+    try:
+        instructor_response, completion = await async_instructor_client.chat.completions.create_with_completion(
+            model=model,
+            api_key=api_key,
+            messages=messages,
+            max_tokens=max_tokens,
+            temperature=temperature,
+            response_model=response_model,
+            timeout=timeout,
+        )
+        response.data = instructor_response.model_dump(by_alias=True)
+    except Exception as e:
+        response.adala_error = True
+        response.adala_message = type(e).__name__
+        response.adala_details = traceback.format_exc()
+    finally:
+        return response.model_dump(by_alias=True)
+
+
+async def parallel_async_get_llm_response(
+    user_prompts: List[str],
+    system_prompt: Optional[str] = None,
+    model: Optional[str] = 'gpt-4o-mini',
+    instruction_first: Optional[bool] = True,
+    response_model: Optional[Type[BaseModel]] = None,
+    api_key: Optional[str] = None,
+    max_tokens: Optional[int] = 1000,
+    temperature: Optional[float] = 0.0,
+    timeout: Optional[Union[float, int]] = None,
+):
+    tasks = [
+        asyncio.ensure_future(
+            async_get_llm_response(
+                user_prompt=user_prompt,
+                system_prompt=system_prompt,
+                model=model,
+                api_key=api_key,
+                max_tokens=max_tokens,
+                temperature=temperature,
+                timeout=timeout,
+                instruction_first=instruction_first,
+                response_model=response_model,
+            )
+        )
+        for user_prompt in user_prompts
+    ]
+    responses = await asyncio.gather(*tasks)
+    return responses
```

Comment:
with httpx, this did not need a separate codepath, you could just pass `concurrent_clients` to the Client from the Runtime like this:
```
            http_client=httpx.AsyncClient(
                limits=httpx.Limits(
                    max_connections=self.concurrent_clients,
                    max_keepalive_connections=self.concurrent_clients,
                ),
                timeout=self.timeout,
            ),
```
is there a way to pass this option through litellm?

Reply:
sort of, you can set the client at the module level (this is what I was doing at first, but decided to get away from it since at the time it seemed we only needed to pass the timeout through the client).
I'm skeptical of assigning the http client at the module level though, it seems error prone... If it works though, I'm all good with it

Reply:
If they are, fantastic, but I've been bitten by that assumption (with libraries that seem focused on being "prod-ready") before